src/snowflake/connector/cursor.py

```diff
@@ -1117,7 +1117,7 @@ def fetch_arrow_batches(self) -> Iterator[Table]:
         )
         return self._result_set._fetch_arrow_batches()
 
-    def fetch_arrow_all(self) -> Table | None:
+    def fetch_arrow_all(self) -> Table:
```

Comment:
the change looks quite neat to me.

@sfc-gh-mkeller , do you think this is a API breaking change that previously we return `None` now we return an empty `Table`.

I'm thinking of introducing a parameter to protect the change

Reply:
**Some thoughts**
1. Here's a suggested fix using your ideas:
```python
from typing import overload, Literal
from warnings import warn
from pyarrow import Table


@overload
def fetch_arrow_all(return_table_even_no_data: Literal[False] = False) -> None:
    ...


@overload
def fetch_arrow_all(return_table_even_no_data: Literal[True] = True) -> Table:
    ...


def fetch_arrow_all(return_table_even_no_data: bool = False) -> Table | None:
    if not return_table_even_no_data:
        warn(
            "The default behaviour of fetch_arrow_all when the query returns zero rows "
            "will change from returning None to returning an empty pyarrow table with "
            "schema using the highest bit length in version X.X.X.",
            DeprecationWarning,
            stacklevel=2,
        )
    # regular code here


foo = fetch_arrow_all(return_table_even_no_data=True) # Table, or "Unknown" if no type stubs for pyarrow are present
bar = fetch_arrow_all(return_table_even_no_data=False) # None

```

I'm happy to change any of this.

2. You _could_ consider this a bug-fix though. It is a discrepancy between the pandas and pyarrow apis. Depends on how strict you want to be about potentially breaking changes.

3. If you follow semver, then I'd expect that you would only introduce breaking changes with major version change. As a user of the connector, it would be nice to know when one expects breaking changes, even if you don't follow semver.

Reply:
hi @thomasaarholt, thanks for the proposal, I personally like it. re your third point, yeah.. sometime I feel torn when facing a breaking change bug fix, but the fact is that there might be programs might depend on this "bug" behavior, so even with a subtle change something might go wrong.

so I feel it would be safer to introduce a parameter for the new behavior.

I'm thinking of a better naming `force_return_table` instead of the `return_table_even_no_data`.

Reply:
can you help make the code change based upon our discussion? really appreciate your contribution and the discussion!

Reply:
Done! I decided to omit the line containing the X.X.X version entirely.